### PR TITLE
Impl improved DocumentTable

### PR DESCRIPTION
**Pull request by ivy-edp**
In this PR, you can see my attempt to incorporate the following functionality (as discussed) into the DocumentTable. I would appreciate feedback:

- Header Redesign: Now resizable for column adjustment (reset with a double-click or component reload).
- Add-Row Button Redesign: Moved below as an Add-Line, no longer within the table (only possible to add one empty line, only visible when no empty row is at the bottom of the table).
- Row Interaction Buttons Redesign: Specific buttons for row interactions appear at the top right when a row is selected (also Delete-Row-Button is placed there).
- Keyboard Interaction Redesign: Tab through rows, and if the last row is not empty it adds a new row and after pressing tap again you should automaticelly be inside this new row.
- Cleanup Table: As soon as you deselect a row by clicking once on the header, the table removes all empty rows from the table. Afterwards, the Add-Row Button appears again.
beispiel_DocumentTable_2

### CODE REVIEW

File: packages/editor/src/components/widgets/table/cell/TableCell.tsx [hunk 1]
```diff
@@ -1,4 +1,8 @@
 import './TableCell.css';
 import type { ReactNode } from 'react';
 
-export const TableCell = (props: { children: ReactNode }) => <td className='table-cell'>{props.children}</td>;
+export const TableCell = (props: { children: ReactNode; cellWidth?: number }) => (
+  <td className='table-cell' style={props.cellWidth ? { width: `${props.cellWidth}px` } : {}}>
+    {props.children}
+  </td>
+);
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/cell/TableCell.tsx:8]: I would export all td properties and set this with on the TableCell directly...

```ts
export const TableCell = ({children, props}: React.HTMLAttributes<HTMLTableCellElement>) => (
  <td {...props} className='table-cell'>
    {children}
  </td>
);
```

```ts
<TableCell key={cell.id} style={{width: cell.column.getSize()}}>
  {flexRender(cell.column.columnDef.cell, cell.getContext())}
</TableCell>
```

File: packages/editor/src/components/widgets/table/footer/TableFooter.tsx [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
+    <div className='add-row-container'>
+      <Button
+        icon={IvyIcons.Plus}
+        onClick={props.addRow}
+        disabled={readonly}
+        aria-label='Add row'
+        onFocus={() => setIsButtonHovered(true)}
+        onBlur={() => setIsButtonHovered(false)}
+        onMouseEnter={() => setIsButtonHovered(true)}
+        onMouseLeave={() => setIsButtonHovered(false)}
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/footer/TableFooter.tsx:25]: why do something via js what css can do too?

File: packages/editor/src/components/widgets/table/footer/TableFooter.css [hunk 3]
```diff
@@ -1,22 +1,51 @@
...
+
+.add-row-container .button:hover {
+  background-color: var(--A300);
+  color: var(--background);
+}
+
+.add-row-container .button:focus {
+  background-color: var(--A300);
+  color: var(--background);
+}
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/footer/TableFooter.css:40]: same styles...

```css
.add-row-container .button:where(:hover, :focus) {
  background-color: var(--A300);
  color: var(--background);
}
```

File: packages/editor/src/components/widgets/table/footer/TableFooter.tsx (hunk 2 above)
Comment (ivy-lli) [reply packages/editor/src/components/widgets/table/footer/TableFooter.tsx:25]: e.g.

```css
.add-row-container .button:hover ~ .add-line {
    background-color: var(--A300);
}
```

File: packages/editor/src/components/widgets/table/header/TableHeader.tsx [hunk 4]
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
 import Button from '../../button/Button';
 import { IvyIcons } from '@axonivy/editor-icons';
 
-export const TableHeader = (props: { colSpan: number; children?: ReactNode }) => (
-  <th className='table-column-header' colSpan={props.colSpan}>
+export const TableHeader = (props: { colSpan: number; children?: ReactNode; columnWidth?: number }) => (
+  <th className='table-column-header' colSpan={props.colSpan} style={props.columnWidth ? { width: `${props.columnWidth}px` } : {}}>
     {props.children}
   </th>
 );
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/header/TableHeader.tsx:11]: same here as on the TableCell

File: packages/editor/src/components/widgets/table/header/TableHeader.tsx [hunk 5]
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
...
-      />
-    </div>
+    <>
+      {props.seperator && (
+        <div
+          {...{
+            onMouseDown: header.header.getResizeHandler(),
+            onTouchStart: header.header.getResizeHandler(),
+            className: `resizer ${header.column.getIsResizing() ? 'isResizing' : ''}`
+          }}
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/header/TableHeader.tsx:23]: why wrap it in an object and then "unpack" it again?
is the following not working? 
```ts
<div onMouseDown={header.header.getResizeHandler()} onTouchStart={header.header.getResizeHandler()} className={`resizer ${header.column.getIsResizing() ? 'isResizing' : ''}`} />
```

File: packages/editor/src/components/widgets/table/row/SelectRow.css [hunk 6]
```diff
@@ -1,14 +1,17 @@
 .selectable-row {
   cursor: pointer;
 }
-.selectable-row:where(:hover) {
+.selectable-row:where(:hover),
+.selectable-row:where(:hover) .input {
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/row/SelectRow.css:5]: where is not needed if there is only one thing inside, just do `.selectable-row:hover`

File: packages/editor/src/components/widgets/table/row/SelectRow.css [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
   background-color: var(--A50);
 }
 
 .selectable-row:where([data-state='selected']),
+.selectable-row:where([data-state='selected']) .table-cell .input,
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/row/SelectRow.css:10]: same here, just do `.selectable-row[data-state='selected']`

File: packages/editor/src/components/widgets/table/header/TableHeader.tsx [hunk 8]
```diff
@@ -4,25 +4,36 @@ import type { ReactNode } from 'react';
...
-      <Button
-        className={header.column.getCanSort() ? 'column-sort-button' : ''}
-        aria-label={`Sort by ${props.name}`}
-        onClick={header.column.getToggleSortingHandler()}
-        data-state={header.column.getIsSorted() || 'unsorted'}
-        icon={header.column.getIsSorted() ? IvyIcons.Chevron : IvyIcons.Straighten}
-      />
-    </div>
+    <>
+      {props.seperator && (
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/header/TableHeader.tsx:17]: a little bit strang that this resizer is the first element in this header and only because of some css it is on the "last" position...
Why isn't this just another item in the flex container? of course then you have to do something else than just `space-between` but that should definitly possible IMO...

File: packages/editor/src/components/widgets/table/header/TableHeader.css [hunk 9]
```diff
@@ -29,3 +39,27 @@
...
+.table-column-header .resizer:hover {
+  width: 5px;
+  background: var(--A300);
+}
+
+.table-column-header .resizer.isResizing {
+  background: var(--A300);
+  width: 5px;
+  opacity: 1;
+}
```
Comment (ivy-lli) [inline packages/editor/src/components/widgets/table/header/TableHeader.css:65]: why is opacity needed?
if this is gone you can combile this styles...

```css
.table-column-header .resizer:where(:hover, .isResizing) {
  width: 5px;
  background: var(--A300);
}
```

File: packages/editor/src/components/widgets/headlineControls/HeadlineControls.css
Comment (ivy-lli) [inline packages/editor/src/components/widgets/headlineControls/HeadlineControls.css:1]: currently this is not used right? So I would revert this change!

File: packages/editor/src/components/widgets/headlineControls/HeadlineControls.tsx
Comment (ivy-lli) [inline packages/editor/src/components/widgets/headlineControls/HeadlineControls.tsx:1]: currently this is not used right? So I would revert this change!

File: packages/editor/src/components/widgets/fieldset/fieldset-control.ts
Comment (ivy-lli) [inline packages/editor/src/components/widgets/fieldset/fieldset-control.ts:1]: currently this is not used right? So I would revert this change!

**Review by ivy-lli — CHANGES_REQUESTED** [review]
Definitely a cool change!

However, I have the feeling that you do not check your changes well enough before adding a reviewer (or at least on this PR)...
I'm aware that you wanted feedback, but I'm a lot faster at review code if its is already cleaned up, e.g. from unused dead prototype code.

File: packages/editor/src/components/widgets/table/footer/TableFooter.tsx (hunk 2 above)
Comment (ivy-edp) [reply packages/editor/src/components/widgets/table/footer/TableFooter.tsx:25]: Oh my God, I've been looking for this specific functionality in CSS, but couldn't find it😂

File: packages/editor/src/components/widgets/table/cell/TableCell.tsx (hunk 1 above)
Comment (ivy-edp) [reply packages/editor/src/components/widgets/table/cell/TableCell.tsx:8]: when i do it like you suggest i get the error: Property 'props' does not exist on type 'HTMLAttributes<HTMLTableCellElement>'

isn't it maybe like this?
`export const TableCell = ({ children, ...props }: React.HTMLAttributes<HTMLTableCellElement>) => (
  <td {...props} className='table-cell'>
    {children}
  </td>
);`

Comment (ivy-lli) [reply packages/editor/src/components/widgets/table/cell/TableCell.tsx]: yes, didn't wrote the code in an ide only here. so probably wrong. Only the idea ;)

**Comment by ivy-lli** [comment]
I'm confused, didn't you want to only do the document table for now? Why did you need also to change the CustomField table? 
If you made a copy of the components instead of changing the existing one you wouldn't have to touch any existing other table or the tests of them...

**Comment by ivy-edp** [comment]
The pull request is now ready for merging:
- Integrated your input from your initial feedback regarding DocumentTable
- Adapted design and usability to other tables (CustomFieldTable, StartCustomFieldTable, MappingTree, ParameterTable)
- Added new Row: SelectableValidationRow (Combination of ValidationRow, MessageRow, SelectRow)
- Adjusted unit tests (UT) and integration tests (IT) for these tables

Open To-Dos:
- MailAttachmentPart, ConditionTable, PropertyTable, TableSort have been adjusted to eliminate UI bugs but still need further refinement in another PR. I have also skipped the tests (UT/IT) for now.
- I actually want to create a generic table component or outsource some parts, as I feel that I have repeated code in many places.

grafik
grafik
grafik
grafik
grafik

**Comment by ivy-lli** [comment]
First thing I already see from the screenshot, in the mapping tree the intention is no longer working.

File: packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx [hunk 10]
```diff
@@ -16,15 +17,53 @@ const useValidationRow = (rowPathSuffix: string | number) => {
...
   const message = useValidationRow(rowPathSuffix);
   return (
-    <MessageRow message={message} title={title}>
+    <MessageRow message={message} title={title} colSpan={colSpan}>
       {children}
     </MessageRow>
   );
 };
 
+export const SelectableValidationRow = <TData extends object>({
```
[the remark this answers is not in the document]
Comment (ivy-lli) [reply packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx:29]: But you can do this also later in another PR.

File: packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx [hunk 11]
```diff
@@ -1,122 +1,162 @@
...
+        }
       }
     }
   });
 
   const action = useAction('openCustomField');
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.value !== '');
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/StartCustomFieldTable.tsx:96]: why not use the empty object?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 12]
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
...
+              key={headerGroup.id}
+              onClick={() => {
+                setRowSelection({});
+              }}
+              onDoubleClick={() => {
+                headerGroup.headers.forEach(header => {
+                  header.column.resetSize();
+                });
+              }}
+            >
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:160]: reuseable component

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 13]
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
...
+            label: 'Open custom field configuration',
+            icon: IvyIcons.GoToSource,
+            action: () => action({ name: table.getRowModel().rowsById[Object.keys(rowSelection)[0]].original.name, type })
+          },
+
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:141]: not on the left side?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 14]
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
...
+              ...data[rowIndex]!,
+              [columnId]: value
+            };
+          }
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' ||
+            updatedData[updatedData.length - 1].type !== 'STRING' ||
+            updatedData[updatedData.length - 1].value !== '') &&
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:103]: why not use the empty object?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 15]
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
...
+        }
       }
     }
   });
 
   const action = useAction('openCustomField');
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.type !== 'STRING' || obj.value !== '');
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:118]: why not use the empty object?

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 16]
```diff
@@ -51,83 +52,132 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
...
+          {table.getRowModel().rows.map(row => (
+            <SelectableValidationRow row={row} colSpan={3} key={row.id} rowPathSuffix={row.original.name}>
+              {row.getVisibleCells().map(cell => (
+                <TableCell key={cell.id}>{flexRender(cell.column.columnDef.cell, cell.getContext())}</TableCell>
+              ))}
+            </SelectableValidationRow>
+          ))}
+        </tbody>
+      </Table>
+      <TableAddRow colSpan={4} addRow={addRow} />
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:179]: is this here always visible?

File: packages/editor/src/components/parts/common/customfield/StartCustomFieldPart.tsx [hunk 17]
```diff
@@ -9,9 +9,9 @@ type StartCustomFieldPartProps = {
 
 const StartCustomFieldPart = ({ customFields, updateCustomFields }: StartCustomFieldPartProps) => {
   return (
-    <PathCollapsible label='Custom Fields' defaultOpen={customFields.length > 0} path='customFields'>
+    <PathContext path='customFields'>
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/customfield/StartCustomFieldPart.tsx:12]: I think you can move this context inside of the customfieldtable? and use there a pathcollapsible?

File: packages/editor/src/components/parts/common/parameter/ParameterTable.tsx [hunk 18]
```diff
@@ -1,108 +1,169 @@
...
+              ...data[rowIndex]!,
+              [columnId]: value
+            };
+          }
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' ||
+            updatedData[updatedData.length - 1].type !== 'String' ||
+            updatedData[updatedData.length - 1].desc !== '') &&
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:91]: why not use the emptyobject?

File: packages/editor/src/components/parts/common/parameter/ParameterTable.tsx [hunk 19]
```diff
@@ -1,108 +1,169 @@
...
+        } else {
+          onChange(updatedData);
+        }
       }
     }
   });
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.type !== 'String' || obj.desc !== '');
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:104]: why not use the empty object?

File: packages/editor/src/components/parts/common/parameter/ParameterTable.tsx [hunk 20]
```diff
@@ -1,108 +1,169 @@
...
+  }, [data, onChange, rowSelection, table]);
+
+  const tableActions =
+    table.getSelectedRowModel().rows.length > 0
+      ? [
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:121]: isn't this on the left side of the row?

File: packages/editor/src/components/parts/common/parameter/ParameterTable.tsx [hunk 21]
```diff
@@ -1,108 +1,169 @@
...
-          </ValidationRow>
-        ))}
-      </tbody>
-      <TableFooter>
-        <TableAddRow colSpan={4} addRow={addRow} />
-      </TableFooter>
-    </Table>
+          </tbody>
+        </Table>
+        {data.filter(obj => obj.name === '' && obj.type === 'String' && obj.desc === '').length === 0 && (
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/common/parameter/ParameterTable.tsx:162]: why not use the empty object?

File: packages/editor/src/components/parts/mail/MailAttachmentPart.tsx [hunk 22]
```diff
@@ -67,32 +67,32 @@ const MailAttachmentTable = () => {
...
+            <ValidationRow colSpan={1} key={row.id} rowPathSuffix={row.index}>
+              {row.getVisibleCells().map(cell => (
+                <TableCell key={cell.id}>{flexRender(cell.column.columnDef.cell, cell.getContext())}</TableCell>
+              ))}
+              <ActionCell actions={[{ label: 'Remove row', icon: IvyIcons.Trash, action: () => removeTableRow(row.index) }]} />
+            </ValidationRow>
+          ))}
+        </tbody>
+      </Table>
+      <TableAddRow colSpan={4} addRow={addRow} />
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/mail/MailAttachmentPart.tsx:95]: always visible?

File: packages/editor/src/components/parts/name/document/DocumentTable.tsx [hunk 23]
```diff
@@ -1,113 +1,167 @@
...
+          if (index === rowIndex) {
+            return {
+              ...data[rowIndex]!,
+              [columnId]: value
+            };
+          }
+          return row;
+        });
+        if (
+          (updatedData[updatedData.length - 1].name !== '' || updatedData[updatedData.length - 1].url !== '') &&
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/name/document/DocumentTable.tsx:85]: why not use the empty object?

File: packages/editor/src/components/parts/name/document/DocumentTable.tsx [hunk 24]
```diff
@@ -1,113 +1,167 @@
...
+        }
       }
     }
   });
 
   const action = useAction('openPage');
 
+  useEffect(() => {
+    if (Object.keys(rowSelection).length !== 1) {
+      const filteredData = data.filter(obj => obj.name !== '' || obj.url !== '');
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/name/document/DocumentTable.tsx:100]: why not use the empty ebject?

File: packages/editor/src/components/parts/name/document/DocumentTable.tsx [hunk 25]
```diff
@@ -1,113 +1,167 @@
...
+          {
+            label: 'Open URL',
+            icon: IvyIcons.GoToSource,
+            action: () => action(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].original.url)
+          },
+          {
+            label: 'Remove row',
+            icon: IvyIcons.Trash,
+            action: () => removeTableRow(table.getRowModel().rowsById[Object.keys(rowSelection)[0]].index)
+          }
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/name/document/DocumentTable.tsx:123]: isn't this on the left side?

File: packages/editor/src/components/parts/name/document/DocumentTable.tsx [hunk 26]
```diff
@@ -1,113 +1,167 @@
...
+              {row.getVisibleCells().map(cell => (
+                <TableCell key={cell.id} style={{ width: cell.column.getSize() }}>
+                  {flexRender(cell.column.columnDef.cell, cell.getContext())}
+                </TableCell>
+              ))}
+            </SelectRow>
+          ))}
+        </tbody>
+      </Table>
+      {data.filter(obj => obj.name === '' && obj.url === '').length === 0 && <TableAddRow colSpan={3} addRow={addRow} />}
```
Comment (ivy-lli) [inline packages/editor/src/components/parts/name/document/DocumentTable.tsx:163]: why not use the empty object?

**Comment by ivy-lli** [comment]
> First thing I already see from the screenshot, in the mapping tree the intention is no longer working.

Broken since 11.2.1. on 11.2.0 it is still working

File: packages/editor/src/components/widgets/table/row/MessageRow.tsx [hunk 27]
```diff
@@ -1,24 +1,26 @@
...
 
-export const MessageRow = forwardRef<HTMLTableRowElement, MessageRowProps>(({ message, children, className, ...props }, forwardRef) => {
-  return (
-    <>
-      <tr ref={forwardRef} className={`row ${message ? `row-${message.severity.toLocaleLowerCase()}` : ''} ${className}`} {...props}>
-        {children}
-      </tr>
-      {message && (
-        <tr className='row row-message'>
-          <td colSpan={100}>
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/components/widgets/table/row/MessageRow.tsx:17]: Yes, I have to set it now because otherwise, the validation message completely disrupts the entire design 😅. However, I have made the field optional now, and if it is not set, I have set the colSpan to 2.

File: packages/editor/src/test-utils/table-utils.tsx [hunk 28]
```diff
@@ -28,54 +28,44 @@ export namespace TableUtil {
...
 
     view.rerender();
-    await assertRowCount(defaultRows + 2);
+    await assertRowCount(expectedRows + 1);
   }
 
   export function assertReadonly() {
     expect(screen.getByRole('button', { name: 'Add row' })).toBeDisabled();
-    expect(screen.getAllByRole('button', { name: 'Remove row' })[0]).toBeDisabled();
+    expect(screen.queryByRole('button', { name: 'Remove row' })).toBeNull();
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/test-utils/table-utils.tsx:69]: Yes, that is correct. However, from what I've seen, it's not possible to select a row in readonly mode. Therefore, I made this comparison directly, but when i think about it, I could also just remove this check. @ivy-lli What do you think?
grafik

File: integrations/standalone/tests/pageobjects/Table.ts [hunk 29]
```diff
@@ -84,7 +84,8 @@ export class Row {
   }
 
   async remove() {
-    await this.locator.getByRole('button', { name: 'Remove row' }).click();
+    await this.locator.click();
+    await this.page.getByRole('button', { name: 'Remove row' }).click();
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply integrations/standalone/tests/pageobjects/Table.ts:88]: Yes, I have also considered that, but unfortunately, it doesn't work with the rowsLocator since this button is completely outside of the table. I wouldn't know how to make it more specific, as the button simply belongs to the collapsible-header div, and there can be multiple of them on a page.

[the remark this answers is not in the document]
Comment (ivy-lli) [reply integrations/standalone/tests/pageobjects/Table.ts:88]: Leave it as it is for now 👍

File: packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx [hunk 30]
```diff
@@ -176,8 +163,8 @@ const CustomFieldTable = ({ data, onChange, type }: CustomFieldTableProps) => {
           ))}
         </tbody>
       </Table>
-      <TableAddRow colSpan={4} addRow={addRow} />
-    </Collapsible>
+      {data.filter(obj => obj === EMPTY_WFCUSTOMFIELD).length === 0 && <TableAddRow addRow={addRow} />}
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:166]: So I am correct in assuming that I should use deepEqual instead?

[the remark this answers is not in the document]
Comment (ivy-lli) [reply packages/editor/src/components/parts/common/customfield/CustomFieldTable.tsx:166]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality

File: packages/editor/src/components/parts/rest/rest-request/rest-body/RestForm.tsx [hunk 31]
```diff
@@ -119,10 +109,8 @@ export const RestForm = () => {
             </ValidationRow>
           ))}
         </tbody>
-        <TableFooter>
-          <TableAddRow colSpan={4} addRow={addRow} />
-        </TableFooter>
       </Table>
+      <TableAddRow addRow={addRow} />
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/components/parts/rest/rest-request/rest-body/RestForm.tsx:113]: Yes, for RestForm, RestParameters, TableSort, MailAttachementPart, and PropertyTable, the 'Add' line is currently always visible, as I haven't really adapted them to the new UX of the tables yet.

File: packages/editor/src/components/widgets/table/row/ReorderRow.css [hunk 32]
```diff
@@ -5,12 +5,12 @@
...
+  i {
+    display: flex;
+    justify-content: flex-end;
+  }
 
-.dnd-row-handle .ivy-change-type {
-  padding: var(--size-2);
+  .ivy-change-type {
+    padding: var(--size-2);
+  }
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/components/widgets/table/row/ReorderRow.css:15]: Oh, I see. Sorry. I will adjust the CSS like this:
```
.dnd-row-handle .ivy-change-type {
  display: flex;
  justify-content: flex-end;
  padding: var(--size-2);
}
```
Because if I were to work with the <i> tag, the 'x' icon of the error message would also be affected, and I deliberately want to only affect the change-type icon with the padding.

[the remark this answers is not in the document]
Comment (ivy-lli) [reply packages/editor/src/components/widgets/table/row/ReorderRow.css:15]: Yes of course this is the other solution :)

File: packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx [hunk 33]
```diff
@@ -37,29 +37,10 @@ export const SelectableValidationRow = <TData extends object>({
...
-      </tr>
-      {message && (
-        <tr className='row row-message'>
-          <td colSpan={colSpan}>
-            <MessageText message={message} />
-          </td>
-        </tr>
-      )}
+      </SelectRow>
+      {message && <MessageRow colSpan={colSpan ? colSpan : 2} message={message} />}
```
[the remark this answers is not in the document]
Comment (ivy-edp) [reply packages/editor/src/components/parts/common/path/validation/ValidationRow.tsx:43]: Oh, yes, you are right. I will revert the change and address this component in the next PR. Sorry.

**Review by ivy-lli — APPROVED** [review]
